Closes #1107: Eliminate outliers by applying thresholds to exploding dimensions of InfoSpace entities

przemyslawjacewicz opened the pull request:
This PR adds the support for eliminating outliers in document metadata. The central point of the implementation is infospace import job `eu.dnetlib.iis.wf.importer.infospace.ImportInformationSpaceJob` and document metadata truncation support in `eu.dnetlib.iis.wf.importer.infospace.truncator.DocumentMetadataAvroTruncator`. For convenient testing additional changes allow to use comprehensible values of truncation parameters.

marekhorst commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/truncator/DocumentMetadataAvroTruncator.java:
```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    public static class TitleTruncator implements Serializable {
```
Do we really need a dedicated Truncator instance for each field? `AbstractTruncator` looks exactly the same as the `TitleTruncator`. 

I am not sure we really need to propagate the truncators model deep down below the `DocumentMetadataAvroTruncator` for each field. I know there are the test classes for each case, which is nice, but maybe we could live with just a truncation test for `DocumentMetadata` and test truncation of all the dimensions. There are only 6 of them. I'd just apply whole truncation logic within the:
```
public DocumentMetadata truncate(DocumentMetadata source)
```
method and drop all the internal truncators. To me it feels overcomplicated.

przemyslawjacewicz replied:
Done, I've removed dedicated truncator classes for each field. The only reason for their implementation was to make unit testing cleaner.

marekhorst commented on iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/infospace/truncator/factory/TestDocumentMetadataAvroTruncatorFactory.java:
```diff
@@ -0,0 +1,17 @@
+
+import eu.dnetlib.iis.wf.importer.infospace.truncator.DocumentMetadataAvroTruncator;
+
+public class TestDocumentMetadataAvroTruncatorFactory implements DocumentMetadataAvroTruncatorFactory {
```
OK, so the truncators solution, apart from being a subject for modular testing, is kind of a wrapper encapsulating the thresholds for 6 dimensions, right? 

So instead of overwriting 6 dimensions for the test we can switch the truncator implementation. It's nice, on one hand because we can keep the workflow definition compact, but on the other hand when having 6 overridable threshold parameters we could make them easily overridable when it turns out our pick statically defined in `ImportInformationSpaceConstants` was not accurate. It might be a matter of a personal opinion but I would prefer to have more flexibility in this regard.

przemyslawjacewicz replied:
After changing the way of passing truncation parameters to importer job this class is no longer needed. Truncation parameters are now passed as dynamic parameters with default values. For tests the values are explicitly provided and there is no need for any 'factory' classes.

marekhorst reviewed (changes requested):
Truncators model is quite interesting but imho quite over-complicated for such a simple truncation operation. I wouldn't even mind truncating the strings/collections directly within the `DocumentMetadataConverter` module, but both aproaches have their pros and cons.

I would leave some of its pieces but at the same time I would change the granularity of `Truncator` usage.

marekhorst commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/truncator/DocumentMetadataAvroTruncator.java:
```diff
@@ -0,0 +1,200 @@
+        public DocumentMetadataAvroTruncator build() {
+            DocumentMetadataAvroTruncator instance = new DocumentMetadataAvroTruncator();
+
+            if (Objects.nonNull(maxAbstractLength)) {
```
If the object is currently instantiated by relying on the statically defined values within `ImportInformationSpaceConstants` class do we need to check the nullability of those fields here and in similar places below?

przemyslawjacewicz replied:
With this implementation the check is necessary. `maxAbstractLength` in this null-check is an `Integer' field within builder nested class. If this field was set using `setMaxAbstractLength()` method then we use it when creating abstract truncator, if this field was not set and was null then we fallback to default abstract truncator with default value created by `DocumentMetadataAvroTruncator` class.

However, builder logic is now changed as a result of changing the way truncation parameters are passed to importer job and this null-check is not needed.  `DocumentMetadataAvroTruncator` is created with explicit parameters provided in the workflow.

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/truncator/factory/DefaultDocumentMetadataAvroTruncatorFactory.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Factory for {@link DocumentMetadataAvroTruncator} providing a default implementation with production values of truncation parameters.
+ */
+public class DefaultDocumentMetadataAvroTruncatorFactory implements DocumentMetadataAvroTruncatorFactory {
```
After changing the way of passing truncation parameters to the importer job this class is no longer needed and I've removed it.

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/resources/eu/dnetlib/iis/wf/importer/infospace/oozie_app/workflow.xml:
```diff
@@ -168,7 +173,8 @@
             <arg>-outputNameOrganization=${output_name_organization}</arg>
             <arg>-outputNameProjectOrganization=${output_name_project_organization}</arg>
-            
+            <arg>-documentMetadataAvroTruncatorFactoryClassName=${document_metadata_avro_truncator_factory_class_name}</arg>
```
After changing the way of passing truncation parameters to importer job this parameter is no longer needed. Instead I've added dynamic parameters with truncation parameters. The values previously set in java code are now default values and are passed to the job with `-D` prefix.

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceConstants.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * {@link eu.dnetlib.iis.importer.schemas.DocumentMetadata} truncation constants.
+ */
+public class ImportInformationSpaceConstants {
```
Done, should've done that at the beginning. This change also results with the following changes:
* `DocumentMetadataAvroTruncator` and `AuthorAvroTruncator` simplification - no nested classes
*  removal of `DocumentMetadataAvroTruncatorFactory` with related classes
* unit and integration test changes

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJob.java:
```diff
@@ -186,16 +192,33 @@ public static void main(String[] args) throws Exception {
+     */
+    private static DocumentMetadataAvroTruncatorFactory documentMetadataAvroTruncatorFactory(String className) {
+        try {
+            return (DocumentMetadataAvroTruncatorFactory) Class.forName(className).getConstructor().newInstance();
```
Done, truncation parameters are now dynamic parameters.

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/truncator/StringTruncator.java:
```diff
@@ -0,0 +1,24 @@
+
+        BreakIterator atWordInstance = BreakIterator.getWordInstance();
+        atWordInstance.setText(source);
+        return source.substring(0, atWordInstance.preceding(length)).trim();
```
Discussed offline: string truncation without word split can potentially remove a large part of source string when the string is short and has a longer word at the truncation index. To avoid this kind of situations we agreed on keeping the word at truncation index. With `BreakIterator` this means using `.following` method and this change is applied here. This will also result in changes in test resources with truncated document metadata.

przemyslawjacewicz commented:
Finished applying changes - moving truncation parameters to workflow params was the most significant change and caused some additional changes.

marekhorst commented on iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/infospace/truncator/DocumentMetadataAvroTruncatorTest.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Test
+    public void shouldTruncateAuthorsWhenItIsNotNull() {
```
I guess we could also add a test verifying whether the number of authors is properly truncated.

przemyslawjacewicz replied:
This unit test checks if authors truncation logic is triggered. It uses a stubbed mock for `Function<List<Author>, List<Author>>` that is responsible for truncation of authors field in `DocumentMetadataAvroTruncator`. So this test method simply checks if the result of truncation is the same as the value used for stubbing. The actual size of the list is not important, that's why I've used a `Collections.singletonList(mock(Author.class))`. The 'real' truncation mechanism is checked in `ImportInformationSpaceJobTest` and in workflow integration test.

marekhorst replied:
OK then. I thought we are missing the unit test proving maxAuthorsSize limitation works as expected.

marekhorst commented on iis-wf/iis-wf-import/src/test/java/eu/dnetlib/iis/wf/importer/infospace/truncator/DocumentMetadataAvroTruncatorTest.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Test
+    public void shouldTruncateKeywordsWhenItIsNotNull() {
```
Same here: I guess we could also add a test verifying whether the number of keywords is properly truncated.

przemyslawjacewicz replied:
And the same answer applies here :)

przemyslawjacewicz commented on iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJob.java:
```diff
@@ -416,6 +457,9 @@ private static String outputPathFor(String path, String subPath) {
         private String outputNameProjectOrganization;
+
+        @DynamicParameter(names = "-D", description = "dynamic parameters related to document metadata truncation")
+        private Map<String, String> documentMetadataTruncationParams = new HashMap<>();
```
The parameters are definitely not dynamic, it would be clearer to define them as normal parameters. I thought that the reason to use dynamic parameters was to keep `ImportInformationSpaceJobParameters` small and clean, but if we can accept a few more fields in this class then I'd go with adding explicit parameters. 

Looking a bit in the future if we find that we need to add even more truncation parameters I'd introduce a dedicated workflow step that would be responsible for truncation. This step would take all the needed parameters as explicit parameters to the job.

przemyslawjacewicz replied:
Or maybe you think we should add a dedicated workflow step now ?

marekhorst replied:
I think we don't need to go that far, I mean I don't expect much more truncation-related parameters to pop-up.

przemyslawjacewicz commented:
> Apart from three rather minor remarks I have one major one which does not seem to be addressed:
> 
> > I am not sure we should bound the truncator with any specific data type because DocumentMetadata is not the only metadata type with fields to be truncated: there is also DataSetReference. Alternatively we would need to introduce dataset truncator handling DataSetReference records.
> 
> Apart from `DocumentMetadata` we also have `DataSetReference` as a candidate for truncation because it holds potentially exploding dimensions:
> 
>     * `creatorNames` (by `max_authors_size` and `max_author_fullname_length`)
> 
>     * `titles` (by `max_titles_size` which should be added and set to `10` by default and and `max_title_length`)
> 
>     * `description` (by `max_description_length`, we could rename already existing `max_abstract_length` and reference same dimension in both places, both are relying on `Result#description` field so the `description` part of the name suits pretty well)
> 
> 
> We've lost it from the radar but I guess we should cover those objects with truncation as well.

I've just added truncation of `DataSetReference` using the description given above. To make truncation a little more generic I've added a method in `ImportInformationSpaceJob` for truncation of any avro type. Apart from that the changes with adding `DataSetReference` are minor.

marekhorst reviewed (approved):
Thanks, LGTM. 

Besides: it turns out we ended up with 1200 added lines, so in the end it wasn't that simple task we initially thought it would be. Probably most of it was related to unit and integration tests but this falls into the bright side ;)

przemyslawjacewicz commented:
Yes, I was also a bit surprised with the overall size of this PR !
